src/common/proxy_wasm/ngx_proxy_wasm_properties.c

```diff
@@ -911,17 +946,27 @@ ngx_proxy_wasm_properties_set_host(ngx_proxy_wasm_ctx_t *pwctx,
     hpn = (host_props_node_t *)
               ngx_str_rbtree_lookup(&pwctx->host_props_tree, path, hash);
 
-    if (value->data == NULL) {
-        if (hpn) {
+    if (hpn) {
+        if (hpn->is_const) {
+            return NGX_DECLINED;
+        }
+
+        ngx_pfree(pwctx->pool, hpn->value.data);
+
+        if (value->data == NULL && !is_const) {
             ngx_rbtree_delete(&pwctx->host_props_tree, &hpn->sn.node);
+
+            return NGX_OK;
         }
 
-        return NGX_OK;
-    }
+        rc = set_hpn_value(hpn, value, is_const, pwctx->pool);
+        if (rc != NGX_OK) {
+            return rc;
+        }
 
-    if (hpn) {
-        ngx_pfree(pwctx->pool, hpn->value.data);
-        new_entry = 0;
+    } else if (value->data == NULL && !is_const) {
+        /* no need to store a non-const NULL */
+        return NGX_OK;
```

Comment:
Can we have this early-exit branch on top as a standalone `if {}`? Overall I find this entire flow **extremely** difficult to read and follow, especially given the APIs aren't symmetrical anymore and `set_host_property` is now an entrypoint (instead of `set_property` only), so I am trying to find ways to make it as clear as possible. Will probably update a lot of it during merge as well...

Reply:
I reworked the flow of the function again!